### PR TITLE
Upgrade to petgraph 0.3

### PR DESCRIPTION
**Pull request by bluss**
Use newest petgraph.

Use their workspace reuse for cycle detection. Also uses a simpler (local) has_path_connecting function for adding a single edge.

It's a good improvement (with keeping the DfsSpace around or not) for add_edge, and a smaller change to add_edges.

Note: Requires Rust 1.12.

Edited: Some background information of the problem at hand in general http://stackoverflow.com/a/20298238/

### CODE REVIEW

**Comment by mitchmindtree** [comment]
Legend! This looks awesome, thanks a lot for the update 😸  Will give this a proper look-over in the morning as I must leave now to stack some Zs ...💤

File: src/lib.rs [hunk 1]
```diff
@@ -105,7 +109,11 @@ impl<N, E, Ix> Dag<N, E, Ix> where Ix: IndexType {
 
     /// Create a new `Dag` with estimated capacity for its node and edge Vecs.
     pub fn with_capacity(nodes: usize, edges: usize) -> Self {
-        Dag { graph: PetGraph::with_capacity(nodes, edges) }
+        let graph = PetGraph::with_capacity(nodes, edges);
+        Dag {
+            graph: graph,
+            cycle_state: DfsSpace::default(),
```
Comment (bluss) [inline src/lib.rs:114]: Oh, the extra local variable graph is a remnant of an earlier version. I'll amend and fix that.

**Comment by mitchmindtree** [comment]
Nice, happy to merge this :+1: I might make some other changes before publishing in order to re-align daggy with petgraph a bit, namely going from iterator to walker using `.detach()` rather than walker to iterator using `.iter()`.

**Comment by bluss** [comment]
Hey, great.

As a warning, maybe a bit frustrating for a dep that's reexporting petgraph, but I think it will have to have multiple breaking releases in a row in the future (no promises on exactly when, though). Mostly to develop graph traits. They are quite tricky and we want to change them around to what works best..
